File: service/internal/platform/github.go
package platform

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/google/go-github/v57/github"
)

type GitHubPlatform struct {
	client        *github.Client
	webhookSecret string
}

type PullRequestEvent struct {
	Owner  string
	Repo   string
	Number int
	Files  []FileChange
}

type FileChange struct {
	Filename string
	Patch    string
	Status   string
}

func NewGitHubPlatform(token, webhookSecret string) *GitHubPlatform {
	client := github.NewClient(nil).WithAuthToken(token)
	return &GitHubPlatform{
		client:        client,
		webhookSecret: webhookSecret,
	}
}

func (g *GitHubPlatform) ValidateWebhook(r *http.Request) error {
	signature := r.Header.Get("X-Hub-Signature-256")
	if signature == "" {
		return fmt.Errorf("missing signature header")
	}

	body, err := io.ReadAll(r.Body)
	if err != nil {
		return err
	}
	r.Body = io.NopCloser(strings.NewReader(string(body)))

	mac := hmac.New(sha256.New, []byte(g.webhookSecret))
	mac.Write(body)
	expectedMAC := "sha256=" + hex.EncodeToString(mac.Sum(nil))

	if !hmac.Equal([]byte(signature), []byte(expectedMAC)) {
		return fmt.Errorf("invalid signature")
	}

	return nil
}

func (g *GitHubPlatform) ParsePullRequest(r *http.Request) (*PullRequestEvent, error) {
	var payload github.PullRequestEvent
	if err := json.NewDecoder(r.Body).Decode(&payload); err != nil {
		return nil, err
	}

	if payload.GetAction() != "opened" && payload.GetAction() != "synchronize" {
		return nil, nil
	}

	pr := payload.GetPullRequest()
	files, _, err := g.client.PullRequests.ListFiles(
		context.Background(),
		payload.Repo.Owner.GetLogin(),
		payload.Repo.GetName(),
		pr.GetNumber(),
		nil,
	)
	if err != nil {
		return nil, err
	}

	var fileChanges []FileChange
	for _, file := range files {
		fileChanges = append(fileChanges, FileChange{
			Filename: file.GetFilename(),
			Patch:    file.GetPatch(),
			Status:   file.GetStatus(),
		})
	}

	return &PullRequestEvent{
		Owner:  payload.Repo.Owner.GetLogin(),
		Repo:   payload.Repo.GetName(),
		Number: pr.GetNumber(),
		Files:  fileChanges,
	}, nil
}

func (g *GitHubPlatform) PostComment(ctx context.Context, owner, repo string, number int, comment string) error {
	_, _, err := g.client.Issues.CreateComment(ctx, owner, repo, number, &github.IssueComment{
		Body: github.String(comment),
	})
	return err
}

File: reviewer/platform_adapters/github.py
import hashlib
import hmac
import json
from typing import Optional

from aiohttp import web
from github import Github

from platform_adapters.models import PullRequestEvent, FileChange


class GitHubPlatform:

    def __init__(self, token: str, webhook_secret: str):
        self.client = Github(token)
        self.webhook_secret = webhook_secret

    async def validate_webhook(self, request: web.Request) -> None:
        signature = request.headers.get("X-Hub-Signature-256")
        if not signature:
            raise ValueError("Missing signature header")

        body = await request.read()
        
        expected_signature = "sha256=" + hmac.new(
            self.webhook_secret.encode(),
            body,
            hashlib.sha256
        ).hexdigest()

        if not hmac.compare_digest(signature, expected_signature):
            raise ValueError("Invalid signature")

    async def parse_pull_request(
        self, request: web.Request
    ) -> Optional[PullRequestEvent]:
        body = await request.read()
        payload = json.loads(body)

        action = payload.get("action")
        if action not in ("opened", "synchronize"):
            return None

        pull_request = payload.get("pull_request", {})
        repository = payload.get("repository", {})
        owner = repository.get("owner", {}).get("login")
        repo_name = repository.get("name")
        pull_request_number = pull_request.get("number")

        repository_object = self.client.get_repo(f"{owner}/{repo_name}")
        pull_request_object = repository_object.get_pull(pull_request_number)
        files = pull_request_object.get_files()

        file_changes = []
        for file in files:
            file_changes.append(FileChange( # type: ignore
                filename=file.filename,
                patch=file.patch or "",
                status=file.status
            ))

        return PullRequestEvent(
            owner=owner,
            repo=repo_name,
            number=pull_request_number,
            files=file_changes # type: ignore
        )

    async def post_comment(
        self, owner: str, repo: str, number: int, comment: str
    ) -> None:
        repository = self.client.get_repo(f"{owner}/{repo}")
        issue = repository.get_issue(number)
        issue.create_comment(comment)